2113 Implementation of Extended Isolation Forest (EIF) anomaly detector

Akhil-Jasson opened the pull request:
This PR implements the Extended Isolation Forest (EIF) algorithm to address [#2113] 
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at our
contribution guide: https://www.aeon-toolkit.org/en/latest/contributing.html.

Feel free to delete sections of this template if they do not apply to your PR,
avoid submitting a blank template or empty sections.
If you are a new contributor, do not delete this template without a suitable
replacement or reason. If in doubt, ask for help. We're here to help!

Please be aware that we are a team of volunteers so patience is
necessary when waiting for a review or reply. There may not be a quick turnaround for
reviews during slow periods. While we value all contributions big or small, pull
requests which do not follow our guidelines may be closed.
-->

#### Reference Issues/PRs

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### What does this implement/fix? Explain your changes.

<!--
A clear and concise description of what you have implemented.
-->

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a dependency, we may suggest adding it as an
optional/soft dependency to keep external dependencies of the core aeon package
to a minimum.
-->

#### Any other comments?

<!--
Any other information that is important to this PR or helpful for reviewers.
-->

### PR checklist

<!--
Please go through the checklist below. Please feel free to remove points if they are
not applicable. To check a box, replace the space inside the square brackets with an
'x' i.e. [x].
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you after the PR has been merged.
- [ ] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.


<!--
Thanks for contributing!
-->

MatthewMiddlehurst commented:
Please fill out the template and use the correct title format.

Akhil-Jasson commented:
I've fixed the title format. Shall I proceed with adding EIF to the online API documentation?

Ramana-Raja commented:
Hi @Akhil-Jasson, I checked out your PR, and the code looks great! That said, I don’t think using H2O is the best approach since Aeon doesn’t rely on it. It might be better to have our own implementation instead.

A few updates to consider:
   1. Could you update the section "Does your contribution introduce a new dependency?" and mention H2O there?

    2.The test cases seem to be missing—could you add them?

    3.Instead of importing the entire H2O module, it’s better to import only what’s needed to keep things lightweight.

MatthewMiddlehurst commented:
New dependencies should be put in pyproject.toml otherwise this won't be tested, Still bits missing from the template

Akhil-Jasson commented:
I've added the h2o dependency to pyproject.toml, but I'm encountering errors when running the test files. The test attempts to import aeon.anomaly_detection._eif but fails, indicating that the module doesn’t exist yet.  

Is there a step I'm missing for adding new modules to aeon? What could be the possible issue?

MatthewMiddlehurst commented:
your import is incorrect.

SebastianSchmidl reviewed (changes requested):
AI2O looks like a massive package. Do we actually want to include it as a dependency? The issue clearly states that we are looking for an implementation in aeon directly.

I know this is not mentioned in the corresponding issue, but I think it makes sense to work with sliding windows in EIF as well. We can always get the original behavior back by setting the window-size to 1.

SebastianSchmidl commented on aeon/anomaly_detection/_eif.py:
```diff
@@ -0,0 +1,331 @@
+        Size of the sliding window. If 1, the original point-wise behavior is used.
+    stride : int, default=1
+        Stride of the sliding window.
+    axis : int, default=1
+        The time point axis of the input series if it is 2D. If ``axis==0``, it is
+        assumed each column is a time series and each row is a time point. i.e. the
+        shape of the data is ``(n_timepoints, n_channels)``. ``axis==1`` indicates
+        the time series are in rows, i.e. the shape of the data is
+        ``(n_channels, n_timepoints)``.
```
Different data layouts are handled in the base class. Please do not expose an additional `axis`-parameter here.

SebastianSchmidl commented on aeon/anomaly_detection/_eif.py:
```diff
@@ -0,0 +1,331 @@
+        ----------
+        X : np.ndarray
+            Training data of shape (n_timepoints,) or (n_timepoints, n_channels)
+        y : np.ndarray, optional
+            Labels for semi-supervised learning. 0 for normal, 1 for anomalous.
+            If None, unsupervised learning is used.
```
Semi-supervised means training on normal data only, so just `0` is allowed here. Otherwise, this would be supervised training with normal data and known anomalies. Details: https://www.aeon-toolkit.org/en/stable/examples/anomaly_detection/anomaly_detection.html#Anomaly-Detection-in-aeon

SebastianSchmidl commented on aeon/anomaly_detection/_eif.py:
```diff
@@ -0,0 +1,331 @@
+        self : object
+            Fitted estimator.
+        """
+        # Ensure X is 2D
+        if X.ndim == 1:
+            X = X.reshape(-1, 1)
```
This should not be necessary anymore if you remove the axis-parameter from the class and use a single data layout within the estimator.

SebastianSchmidl commented on aeon/anomaly_detection/_eif.py:
```diff
@@ -0,0 +1,331 @@
+            tree = self._build_tree(X_sample, rng)
+            self.forest_.append(tree)
+
+        # Calculate anomaly scores
+        scores = self._predict(X)
+
+        # Set threshold
+        self.threshold_ = float(np.percentile(scores, 100 * (1 - self.contamination)))
```
This looks suspicious. Why do you predict in `fit`?

SebastianSchmidl commented on aeon/anomaly_detection/_eif.py:
```diff
@@ -0,0 +1,331 @@
+                scores,
+                window_size=self.window_size,
+                stride=self.stride,
+                n_timepoints=self._n_samples_orig,
```
This just works if either EIF is used in an unsupervised way or both training and prediction time series have the same length!

SebastianSchmidl commented on aeon/anomaly_detection/_eif.py:
```diff
@@ -0,0 +1,331 @@
+            return 0
+        elif n == 2:
+            return 1
+        return 2 * (np.log(n - 1) + 0.5772156649) - 2 * (n - 1) / n
```
Where is the magic number from? Please at least add a comment.

SebastianSchmidl reviewed (changes requested):
The code does not look too complicated, so I think, it was a good decision to not include H2O.ai as a dependency!

Thanks for the implementation effort. Please address the issues below and then compare your implementation to the original implementation by Hariri et al. to demonstrate that your/our implementation produces the same results as theirs.

I think this estimator is a suitable candidate for improving the performance with Numba / JITs. But let us first focus on making the implementation correct, and then tackle the performance optimization in another PR.

Akhil-Jasson commented:
Previously, the tests were triggered automatically on commit based on the old repo structure. Since then, the structure has changed significantly. I ran the tests locally using the earlier setup, and they passed.

To accommodate the requirements of the test files—such as the need for y in the fit function—I’ve made it an optional parameter. I’ll take a look into the other comments as well.

Given the extent of changes in the repo structure, is it easier to just create a new branch from the updated main branch or is there an easier alternative? Please let me know!

MatthewMiddlehurst commented:
I would merge main into your branch, the bulk of your code should not conflict. It is mainly that we have refactored the module so it has all moved around a bit. If you do not know which module to put your implementation in just put it anywhere for now and we can discuss after where is best.

Akhil-Jasson commented:
Changes have been made to this branch to align with the newly refactored module structure. I'll soon address the feedback to remove the axis hyperparameter and add the necessary comments as requested.

MatthewMiddlehurst commented:
Hello, is this still active?

Akhil-Jasson commented:
Hello @MatthewMiddlehurst,
Thank you for checking in. I apologize for the long pause. I've had some time away, but I’m still eager to continue work on the Extended Isolation Forest implementation.

Akhil-Jasson commented:
Hi @SebastianSchmidl  and @MatthewMiddlehurst,

I've addressed the feedback from the previous review:

Key Changes Made:
- Removed threshold calculation from _fit()
- Fixed sliding window length handling - Now works with different-sized prediction data (addresses the length dependency issue)
- Added comments for magic numbers - Explained the Euler-Mascheroni constant in _c() method
- Maintained unsupervised/semi-supervised support - Works with both `y=None` and `y=0` modes

Please let me know if there are any remaining concerns.

Thanks for the detailed feedback, it really helped improve the implementation!

MatthewMiddlehurst commented on aeon/anomaly_detection/outlier_detection/_eif.py:
```diff
@@ -0,0 +1,347 @@
+    random hyperplanes for splitting instead of axis-parallel splits.
+    This allows for better handling of high-dimensional data and complex distributions.
+
+    The implementation supports both unsupervised and semi-supervised learning:
+    - Unsupervised: No labels provided (y=None)
+    - Semi-supervised: Labels provided (y=0 for normal data)
```
by unsupervised we mean it makes predictions on the same data with no labels, for semi-supervised it fits on data then this can be used for other data. I do not think this is unsupervised in our framework.

MatthewMiddlehurst commented on aeon/anomaly_detection/outlier_detection/_eif.py:
```diff
@@ -0,0 +1,347 @@
+        # Determine sample size
+        n_samples = X_windowed.shape[0]
+        if isinstance(self.sample_size, str):
+            if self.sample_size != "auto":
+                raise ValueError("sample_size must be 'auto', int or float")
```
this error catching could be better. it will only run if it is an incorrect str

MatthewMiddlehurst commented on aeon/anomaly_detection/outlier_detection/_eif.py:
```diff
@@ -0,0 +1,347 @@
+        else:
+            X_windowed = X
+
+        # Set random state
+        rng = np.random.RandomState(self.random_state)
```
can you use https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html for rng, see the docstring of other algorithms which use rng for a better one

TonyBagnall commented:
hi @Akhil-Jasson in the spirit of housekeeping I am converting this to draft, feel free to reopen if you want to continue with it
